File: src/main/java/bad/robot/radiate/ui/Transparency.java
package bad.robot.radiate.ui;

import bad.robot.http.configuration.AbstractValueType;

import java.awt.*;

import static java.awt.AlphaComposite.SRC_OVER;


public class Transparency extends AbstractValueType<Float> {

    public static Transparency Transparent = new Transparency(0.0f);
    public static Transparency TwentyPercent = new Transparency(0.20f);
    public static Transparency SeventyFivePercent = new Transparency(0.75f);
    public static Transparency Opaque = new Transparency(1.0f);

    public Transparency(float percentage) {
        super(percentage);
    }

    public Transparency increase(float percentage) {
        return new Transparency(value + percentage);
    }

    /**
     * Create a {@link java.awt.AlphaComposite} with transparency set on top of the existing composite. Set the transparency
     * relative to the current transparency (ie, 50% of the already 50% transparency).
     */
    public AlphaComposite createAlphaComposite(AlphaComposite current) {
        return AlphaComposite.getInstance(SRC_OVER, current.getAlpha() * value);
    }

    public AlphaComposite createAlphaComposite() {
        return AlphaComposite.getInstance(SRC_OVER, value);
    }

}

File: src/main/scala/bad/robot/radiate/ui/Transparency.scala
package bad.robot.radiate.ui

import java.awt.AlphaComposite.SRC_OVER
import java.awt._

object Transparency {
  var Transparent = Transparency(0.0f)
  var TwentyPercent = Transparency(0.20f)
  var SeventyFivePercent = Transparency(0.75f)
  var Opaque = Transparency(1.0f)
}

case class Transparency(percentage: Float) {

  def +(percentage: Float) = increase(percentage)

  def increase(percentage: Float): Transparency = {
    Transparency(this.percentage + percentage)
  }

  /**
   * Create a [[java.awt.AlphaComposite]] with transparency set on top of the existing composite. Set the transparency
   * relative to the current transparency (ie, 50% of the already 50% transparency).
   */
  def createAlphaComposite(current: AlphaComposite): AlphaComposite = {
    AlphaComposite.getInstance(SRC_OVER, current.getAlpha * this.percentage)
  }

  def createAlphaComposite: AlphaComposite = {
    AlphaComposite.getInstance(SRC_OVER, this.percentage)
  }
}